streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/HighAvailabilityTaskAssignor.java

```diff
@@ -0,0 +1,544 @@
+                if (destination == null) {
+                    log.error("Task {} is assigned to client {} in initial assignment but has no owner in the final " +
+                                  "balanced assignment.", task, source);
+                    throw new IllegalStateException("Found task in initial assignment that was not assigned in the final.");
+                } else if (source != destination) {
+                    movements.add(new Movement<>(task, source, destination));
+                }
+            }
+        }
+        return movements;
+    }
+
+    /**
+     * @return true iff all active tasks with caught-up client are assigned to one of them, and all tasks are assigned
+     */
+    boolean previousAssignmentIsValid() {
+        final Set<TaskId> unassignedActiveTasks = new HashSet<>(allTasks);
+        final Map<TaskId, Integer> unassignedStandbyTasks =
+            configs.numStandbyReplicas == 0 ?
+                Collections.emptyMap() :
+                new HashMap<>(statefulTasksToRankedCandidates.keySet().stream()
+                                  .collect(Collectors.toMap(task -> task, task -> configs.numStandbyReplicas)));
+
+        for (final Map.Entry<ID, ClientState> clientEntry : clientStates.entrySet()) {
+            final ID client = clientEntry.getKey();
+            final ClientState state = clientEntry.getValue();
+            final Set<TaskId> prevActiveTasks = state.prevActiveTasks();
+
+            // Verify that this client was caught-up on all stateful active tasks
+            for (final TaskId activeTask : prevActiveTasks) {
+                if (!taskIsCaughtUpOnClient(activeTask, client)) {
+                    return false;
+                }
+            }
+            unassignedActiveTasks.removeAll(prevActiveTasks);
+
+            if (!unassignedStandbyTasks.isEmpty()) {
+                for (final TaskId task : state.prevStandbyTasks()) {
+                    final Integer remainingStandbys = unassignedStandbyTasks.get(task);
+                    if (remainingStandbys != null) {
+                        if (remainingStandbys == 1) {
+                            unassignedStandbyTasks.remove(task);
+                        } else {
+                            unassignedStandbyTasks.put(task, remainingStandbys - 1);
+                        }
+                    }
+                }
+            }
+        }
+        return unassignedActiveTasks.isEmpty() && unassignedStandbyTasks.isEmpty();
+    }
+
+    /**
+     * @return true if this client is caught-up for this task, or the task has no caught-up clients
+     */
+    boolean taskIsCaughtUpOnClient(final TaskId task, final ID client) {
+        boolean hasNoCaughtUpClients = true;
+        final SortedSet<RankedClient<ID>> rankedClients = statefulTasksToRankedCandidates.get(task);
+        if (rankedClients == null) {
+            return true;
+        }
+        for (final RankedClient rankedClient : rankedClients) {
```

Comment:
Please avoid raw types, even when you don't need the type bound.

```suggestion
        for (final RankedClient<ID> rankedClient : rankedClients) {
```

Reply:
IDEA will now yell at me for this so you don't have to 🙂